tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/effects/debugger_effects.ts

```diff
@@ -62,9 +139,152 @@ export class DebuggerEffects {
     )
   );
 
+  /** @export */
+  readonly loadNumExecutions$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(debuggerRunsLoaded),
+      withLatestFrom(this.store.select(getNumExecutionsLoaded)),
+      filter(([props, loaded]) => {
+        return (
+          Object.keys(props.runs).length > 0 &&
+          loaded.state !== DataLoadState.LOADING
+        );
+      }),
+      tap(() => this.store.dispatch(numExecutionsRequested())),
+      mergeMap(([props, loaded]) => {
+        // TODO(cais): Handle multple runs. Currently it is assumed that there
+        // is at most only one debugger run available.
+        const runId = Object.keys(props.runs)[0];
+        const begin = 0;
+        const end = 0;
+        return this.dataSource.fetchExecutionDigests(runId, begin, end).pipe(
+          map((digests) => {
+            return numExecutionsLoaded({numExecutions: digests.num_digests});
+          })
+        );
+        // TODO(cais): Add catchError() to pipe.
+      })
+    )
+  );
+
+  /** @export */
+  readonly initialExecutionDigestsLoading$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(numExecutionsLoaded),
+      withLatestFrom(
+        this.store.select(getActiveRunId),
+        this.store.select(getDisplayCount),
+        this.store.select(getExecutionDigestsLoaded)
+      ),
+      filter(([props, runId, _, loaded]) => {
+        return (
+          props.numExecutions > 0 &&
+          runId !== null &&
+          Object.keys(loaded.pageLoadedSizes).length === 0 &&
+          loaded.state !== DataLoadState.LOADING
+        );
+      }),
+      tap(() => this.store.dispatch(executionDigestsRequested())),
+      mergeMap(([props, runId, displayCount, _]) => {
+        const begin = 0;
+        const end = Math.min(props.numExecutions, displayCount);
+        return this.dataSource.fetchExecutionDigests(runId!, begin, end).pipe(
+          map((digests) => {
+            return executionDigestsLoaded(digests);
+          })
+        );
+        // TODO(cais): Add catchError() to pipe.
+      })
+    )
+  );
+
+  /** @export */
+  readonly loadExecutionDigests$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(requestExecutionDigests),
+      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
+      filter(([_, loaded]) => loaded.state !== DataLoadState.LOADING),
+      map(([props, loaded]) => {
+        return {
+          props,
+          loaded,
+          missingPages: getMissingPages(
+            props.begin,
+            props.end,
+            props.pageSize,
+            loaded.numExecutions,
+            loaded.pageLoadedSizes
+          ),
+        };
+      }),
+      filter(({missingPages}) => missingPages.length > 0),
+      tap(() => this.store.dispatch(executionDigestsRequested())),
+      withLatestFrom(this.store.select(getExecutionDigestsLoaded)),
```

Comment:
Please just consider using `loaded` from L205 instead. That would make argument to mergeMap simpler, too.

Reply:
Done. Thanks for catching this.